File: app/javascript/channels/appearance_channel.js
import consumer from "./consumer"

consumer.subscriptions.create({channel: "AppearanceChannel"}, {
  initialized() {
    this.update = this.update.bind(this)
  },

  connected() {
    this.install()
    this.update()
  },

  disconnected() {
    this.uninstall()
  },

  rejected() {
    this.uninstall()
  },

  update() {
    this.documentIsActive ? this.update_online_status() : this.away()
  },

  update_online_status() {
    this.perform("update_online_status")
  },

  away() {
    this.perform("away")
    this.uninstall()
  },

  received(data) {
    let { user } = data;
    let { nickname, status, id } = user;
    let userNodeId = $('#user-list').find(`[data-user-id=${id}]`);

    if (!status){
      userNodeId.remove();
    };

    if (status && userNodeId.length != 1){
      $('#user-list').append(`
        <li data-user-id="${id}" data-user-status="${status}">
          <b>${nickname}</b>
          <span class="spinner-grow spinner-grow-sm text-success">&#183;</span>
        </li>`);
    };
  },

  install() {
    window.addEventListener("focus", this.update)
    window.addEventListener("blur", this.update)
    document.addEventListener("turbolinks:load", this.update)
    document.addEventListener("visibilitychange", this.update)
  },

  uninstall() {
    window.removeEventListener("focus", this.update)
    window.removeEventListener("blur", this.update)
    document.removeEventListener("turbolinks:load", this.update)
    document.removeEventListener("visibilitychange", this.update)
  },

  get documentIsActive() {
    return document.visibilityState == "visible" && document.hasFocus()
  },
});

File: app/channels/appearance_channel.rb
class AppearanceChannel < ApplicationCable::Channel
  def subscribed
    stream_from "appearance_channel"

    current_user_appear

    update_online_status
  end

  def unsubscribed
    current_user_away

    update_online_status
  end

  def away
    current_user_away

    update_online_status
  end

  def update_online_status
    ActionCable.server.broadcast "appearance_channel", user: current_user
  end

  private

  def current_user_away
    current_user.update!(status: false)
  end


  def current_user_appear
    current_user.update!(status: true)
  end
end
